lib/throttle-stream.js

```diff
@@ -0,0 +1,173 @@
+ */
+function ThrottleStream (opts) {
+  if (!new.target) {
+    return new ThrottleStream(opts)
+  }
+
+  Transform.call(this, opts)
+
+  /**
+   * The function to calculate bytes per second to throttle to
+   * @type {Function}
+   * @public
+   * @memberof ThrottleStream
+   * @name bytesPerSecondFn
+   * @default null
+   */
+  this.bytesPerSecondFn = opts?.bytesPerSecond === undefined
+    ? function bytesPerSecondFn (_elapsedTime, _bytes) {
+      return 16384
+    }
+    : typeof opts.bytesPerSecond === 'number'
+      ? function bytesPerSecondFn (_elapsedTime, _bytes) {
+        return opts.bytesPerSecond
+      }
+      : opts.bytesPerSecond
+
+  /**
+   * The time that the stream started
+   * @type {number}
+   * @public
+   * @memberof ThrottleStream
+   * @name startTime
+   * @default null
+   */
+  this.startTime = null
+
+  /**
+   * The number of bytes that have been sent through the stream
+   * @type {number}
+   * @public
+   * @memberof ThrottleStream
+   * @name bytes
+   * @default 0
+   */
+  this.bytes = 0
+
+  /**
+   * The buffered chunk.
+   * @type {Buffer}
+   * @private
+   * @memberof ThrottleStream
+   * @name _buffer
+   * @default null
+   */
+  this._buffer = null
+
+  this._interval = null
+
+  /**
+   * Number of bytes allowed to be pushed to the next stream
+   * @type {number}
+   * @private
+   * @default 0
+   */
+  this._allowedBytes = 0
+
+  this._callback = null
+}
+
+ThrottleStream.prototype = Object.create(Transform.prototype, {
+  constructor: {
+    value: ThrottleStream,
+    enumerable: false,
+    writable: true,
+    configurable: true
+  }
+})
+
+/**
+ * The internal `_transform` function is called by the `write` function of the
+ * `Transform` stream base class. This function is not intended to be called
+ * directly.
+ * @param {Buffer} chunk
+ * @param {string} encoding
+ * @param {Function} fn
+ */
+ThrottleStream.prototype._transform = function _transform (chunk, encoding, callback) {
+  chunk && (this._buffer = chunk)
+  callback && (this._callback = callback)
+
+  try {
+    this.startTime === null && this._init()
+
+    if (this._allowedBytes !== 0) {
+      const chunkLength = this._buffer.length
+      if (chunkLength <= this._allowedBytes) {
+        this._allowedBytes -= chunkLength
+        this.push(this._buffer)
+        this._buffer = null
+        this._callback()
+        return
+      }
+
+      const allowedBytes = this._allowedBytes
+      this._allowedBytes = 0
+      this.push(this._buffer.subarray(0, allowedBytes))
+      this._buffer = this._buffer.subarray(allowedBytes)
+    }
+  } catch (err) {
+    // Pass the error to the callback to handle it in the stream
+    this._callback(err)
+  }
+}
+
+ThrottleStream.prototype._init = function _init () {
+  try {
+    this.startTime = Date.now()
+
+    this._allowedBytes = this.bytesPerSecondFn(0, 0)
+    this._interval = setInterval(intervalHandler.bind(this), 1000)
```

Comment:
You mean like an observable?

Reply:
Thought about it. I dont know if is worth. Instead having multiple intervals we need to keep references on the streams. We then need to register the stream on creation to a global inverval, and deregister on destroy. 
And in the global Interval we have to iterate over all array elements and call the intervalhandler. Is that much better than having separated local intervals?

Reply:
Have you considered maybe not using `setInterval` but just relying on the `setImmediate`?
It might be more cumbersome but as you are picking from the buffer every time the network has been flushed for that request(socket), it might be some sort of optimistic flush as you just calculate the delta between timestamps and flush it right away with another `setImmediate` schedule.

For a more _fine grained_ management of chunks, it will require to implement manually the duplex, beyond only the `transform`. And in this state should be a good way to go already

Reply:
> Is that much better than having separated local intervals?

Don't ask me for a bench, but I did it in the past to solve some prod issue/optimization due GC slowing down the system.

But as said: i think it can be a follow-up PR IMHO the current work is good enough!